### ISSUE TITLE
Feedback: 'Asynchronous Handlers'

### ISSUE THREAD

**Issue by cremor** [Particular/docs.particular.net#2693]
Feedback for 'Asynchronous Handlers' https://docs.particular.net/nservicebus/handlers/async-handlers

Location in GitHub: https://github.com/Particular/docs.particular.net/blob/master/nservicebus/handlers/async-handlers.md

ping @Particular/nservicebus-maintainers 

I generally like this documentation page very much. It shows that you not only want to document your code, but also help the customer to use it better and understand the reasons behind it. :thumbsup:

But having said this, I have a few issues/questions about that page:

### (A) Handlers in IO thread pool thread
In the "Introduction" section it says

> Handlers and Sagas will be invoked from an IO thread pool thread.

It this really true? Or to be more precise, is it true for all transports? Because it seems like MSMQ calls the handlers in a `Task.Run` which is - as explained just a few paragraphs down - a worker thread pool thread. Relevant code for MSMQ: https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Transports/Msmq/MessagePump.cs#L175

### (B) Multiple thread pools
Are there even two separate thread pools? I am for sure no expert in this topic, but I've now spend some hours reading documentation and found the following:

From a native (Windows) point of view it seems like there is no distinction any more since Windows Vista. From the MSDN article [Thread Pools](https://msdn.microsoft.com/en-us/library/windows/desktop/ms686760.aspx) you linked under "Further reading":

> The original thread pool has been completely rearchitected in Windows Vista. The new thread pool is improved because it provides a single worker thread type (supports both I/O and non-I/O) [...]

From a managed (.NET) point of view the article [The Managed Thread Pool](https://msdn.microsoft.com/en-us/library/0ka9477y.aspx) says:

> There is only one thread pool per process.

Is there some official documentation that mentions these two thread pools?

### (C) ConfigureAwait
I fully understand why you - as a library developer - need to use `ConfigureAwait(false)` everywhere. But according to the section [Usage of ConfigureAwait](https://docs.particular.net/nservicebus/handlers/async-handlers#usage-of-configureawait) we should also specify `ConfigureAwait(false)` in all message handlers. Could you please explain why this is? Since the message handlers are called from a thread pool thread they shouldn't have a `SynchronizationContext` set anyway, right? So what context is there that will be restored after an `await`?

**Comment by danielmarbach** [comment]
@Particular/nservicebus-maintainers will deal with this tomorrow

**Comment by danielmarbach** [comment]
## Handlers in IO thread pool thread

You are right, it is not really true for all transports. In fact, it is also highly dependent on the APIs that are called. So, in general, this statement is an oversimplification. As you mentioned with MSMQ because of the synchronous nature of the APIs we explicitly schedule on a thread from the worker pool. Depending on the implementation of the client that drives the transport this can vary. Data can be pushed from sockets, network IO etc. directly into NSB since the transport is in full control. It is still possible that the driver could be fully IO bound and then the statement still holds up. The problem is more that for example many of the BCL implementations (i.ex. HttpClient) schedule continuations on the worker pool because, stating David Fowler: The user cannot be trusted ;) So any await statement somewhere inside the handling pipelines will influence it.  I thought the simplication is good enough. Any recommendations how to rephrase it to make it less confusing?

## Multiple Thread pools

It is a bit of a wording issue. Depending what you read or where you are coming from you'll find people mentioning the following:

* A thread pool with two different pools of threads (worker vs IO)
* Two thread pools

> Specifically, it’s worth mentioning that there are two independent implementations of the ThreadPool: one deals with CPU parallelism and is referred to as the worker ThreadPool; the other deals with I/O Parallelism and can be dubbed I/O ThreadPool [Link](https://msdn.microsoft.com/en-us/magazine/ff960958.aspx)

> (Really there are even more ways to queue work.  As mentioned in my previous post we really have two pools of threads – the “worker threads” and the “I/O threads.”  Work is queued to the I/O threads in response to the completion of asynchronous I/O, or manually via ThreadPool.UnsafeQueueNativeOverlapped. [Link](https://blogs.msdn.microsoft.com/ericeil/2009/04/23/clr-4-0-threadpool-improvements-part-1/)

Interesting article and discussion by Stephen Cleary

> Yes, it's one of the I/O threads in the thread pool. The ThreadPool keeps a number of threads registered in its IOCP; these are different than the worker threads that most people associate with the ThreadPool. The ThreadPool manages both worker thread and I/O threads.

http://blog.stephencleary.com/2013/11/there-is-no-thread.html

But based on your feedback I will reword it to two pools of threads and mention one thread pool if that helps to avoid confusion

## ConfigureAwait

It is not strictly necessary to have `ConfigureAwait(false)`. You are absolutely right there. Under the assumption, there is nothing to restore (and that holds up true most of the time) adding the `ConfigureAwait(false)` "only" saves you [a few extract conditions](https://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/Task.cs,2974) to check in the CRL. But there is another caveat. By not explicitely adding `ConfigureAwait(false) `you are essentially tying yourself into assumptions how your code will be executed. It could be perfectly valid that the client of your transport in a new major version introduces a scope that is flowed. Or if you switch transports another transport could have it (currently not the case though so might be a weak argument). In the end it is your code and your decision. We just felt it would be way to nuanced if we write down the explanation above and then hold our fingers crossed that developers get it right.

**Comment by cremor** [comment]
Wow, thanks for the detailed answer!

### (A) Handlers in IO thread pool thread
> Any recommendations how to rephrase it to make it less confusing?

I would just mention that it isn't 100% sure which type of thread executes the handler. Something like "Handlers and Sagas will be invoked from a thread pool thread. Depending on the transport implementation it might use a worker thread pool thread or an IO thread pool thread.".

A later section explains how long running CPU bound code should be written and already contains a big warning about analyzing your own code. Maybe that could be a bit expanded to mention that this is only needed if the handler is called in an IO thread pool thread. But that could also lead to further questions, because as far as I know, there is no way to check which type of thread you are currently in. So maybe not changing this is better.

### (B) Multiple thread pools
I only looked over that links quickly, but will read it later. Thanks!

### (C) ConfigureAwait
Understood, thanks.

Maybe this is offtopic for this docs discussion, but did you think about calling the message handlers explicitly without a current `SynchronizationContext` when you implemented v6? Something like shown in the blog post [An alternative to ConfigureAwait(false) everywhere](https://blogs.msdn.microsoft.com/benwilli/2017/02/09/an-alternative-to-configureawaitfalse-everywhere/) or the [SynchronizationContextSwitcher from the AsyncEx library](https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Tasks/SynchronizationContextSwitcher.cs#L36) which was linked by Stephen Cleary in this blog post.

But then again, as you said, in theory a transport could introduce its own `SynchronizationContext` and using such an alternative would make it impossible for the user to use this context. So I assume while you suggest to not flow the context in message handlers, you also don't want to make it impossible for your customers?

**Comment by danielmarbach** [comment]
@cremor I sent in https://github.com/Particular/docs.particular.net/pull/2697 Happy to hear your review over there if you don't mind

## (A) Handlers in IO thread pool thread

see PR for more details. I think I'll leave the rest as is

## (C) ConfigureAwait

Yes we did think about it. We decided against magic in the framework and stick to current async best practices in order to benefit from the depth of knowledge and "wisdom" that is already available and taught in many courses, presentation etc. 

> So I assume while you suggest to not flow the context in message handlers, you also don't want to make it impossible for your customers?

Correct 💃

**Comment by cremor** [comment]
Just for the fun, here is something I found while going through articles linked in the articles you linked:

[Windows I/O threads vs. managed I/O threads](https://blogs.msdn.microsoft.com/ericeil/2008/06/20/windows-io-threads-vs-managed-io-threads/)

> In the managed ThreadPool, we use the terms “worker thread” and “I/O thread.”  In our case, an I/O thread is one that waits on a completion port; i.e., it’s exactly equivalent to Windows’ non-I/O thread.  How confusing!

Good job Microsoft!
